bug: prometheus causing crash during rewards generation

Reported by a user on the community slack channel:

Config:
```yaml
args:
  - run
  - --ethereum.use_native_batch_call=false
  - --ethereum.native_batch_call_size=10
  - --prometheus.enabled=true
  - --prometheus.port=9092

config:
  SIDECAR_DEBUG: "false"
  SIDECAR_CHAIN: "mainnet"
[skipped DB settings]
```

Logs:
```
{"level":"info","ts":"2025-03-04T12:27:29.111Z","caller":"pipeline/pipeline.go:287","msg":"Calculating rewards for snapshot date","cutoffDate":"2024-10-14","rewardsCalculationEnd":"2024-10-13","blockNumber":20972916}
{"level":"info","ts":"2025-03-04T12:27:29.111Z","caller":"rewardsCalculatorQueue/queue.go:23","msg":"Enqueueing rewards calculation message","data":{"CalculationType":"calculateRewards","CutoffDate":"2024-10-14"}}
{"level":"info","ts":"2025-03-04T12:27:29.111Z","caller":"rewardsCalculatorQueue/queue.go:37","msg":"Waiting for rewards calculation response","data":{"CalculationType":"calculateRewards","CutoffDate":"2024-10-14"}}
{"level":"info","ts":"2025-03-04T12:27:29.111Z","caller":"rewardsCalculatorQueue/handler.go:12","msg":"Processing rewards calculation message","data":{"CalculationType":"calculateRewards","CutoffDate":"2024-10-14"}}
{"level":"info","ts":"2025-03-04T12:27:29.111Z","caller":"rewards/rewards.go:98","msg":"Acquired rewards generation lock","snapshotDate":"2024-10-14"}
{"level":"info","ts":"2025-03-04T12:27:29.116Z","caller":"rewards/rewards.go:150","msg":"Calculating rewards for snapshot date","snapshot_date":"2024-10-14"}
{"level":"info","ts":"2025-03-04T12:32:58.288Z","caller":"sidecar/blockIndexer.go:265","msg":"New tip found, updating","latestTip":21973514,"currentTip":21973482}
{"level":"info","ts":"2025-03-04T12:34:02.267Z","caller":"rewards/1_goldActiveRewards.go:109","msg":"Generating active rewards","rewardsStart":"1970-01-01 00:00:00","cutoffDate":"2024-10-14","destTableName":"gold_1_active_rewards_2024_10_14"}
{"level":"info","ts":"2025-03-04T12:37:32.044Z","caller":"rewards/2_goldStakerRewardAmounts.go:153","msg":"Generating staker reward amounts","cutoffDate":"2024-10-14","destTableName":"gold_2_staker_reward_amounts_2024_10_14","amazonHardforkDate":"2024-08-02","nileHardforkDate":"2024-08-12","arnoHardforkDate":"2025-01-21","trinityHardforkDate":"2025-01-21"}
{"level":"info","ts":"2025-03-04T12:38:59.488Z","caller":"sidecar/blockIndexer.go:265","msg":"New tip found, updating","latestTip":21973546,"currentTip":21973514}
{"level":"info","ts":"2025-03-04T12:45:31.005Z","caller":"sidecar/blockIndexer.go:265","msg":"New tip found, updating","latestTip":21973578,"currentTip":21973546}
{"level":"info","ts":"2025-03-04T12:52:03.166Z","caller":"sidecar/blockIndexer.go:265","msg":"New tip found, updating","latestTip":21973610,"currentTip":21973578}
{"level":"info","ts":"2025-03-04T12:58:35.086Z","caller":"sidecar/blockIndexer.go:265","msg":"New tip found, updating","latestTip":21973642,"currentTip":21973610}
{"level":"info","ts":"2025-03-04T13:04:37.337Z","caller":"sidecar/blockIndexer.go:265","msg":"New tip found, updating","latestTip":21973674,"currentTip":21973642}
{"level":"info","ts":"2025-03-04T13:10:36.733Z","caller":"rewards/3_goldOperatorRewardAmounts.go:43","msg":"Generating staker reward amounts","cutoffDate":"2024-10-14","destTableName":"gold_3_operator_reward_amounts_2024_10_14"}
{"level":"info","ts":"2025-03-04T13:10:46.570Z","caller":"rewards/4_goldRewardsForAll.go:74","msg":"Generating rewards for all table","cutoffDate":"2024-10-14","destTableName":"gold_4_rewards_for_all_2024_10_14"}
{"level":"info","ts":"2025-03-04T13:10:47.606Z","caller":"rewards/5_goldRfaeStakers.go:144","msg":"Generating rfae stakers table","cutoffDate":"2024-10-14","destTableName":"gold_5_rfae_stakers_2024_10_14","arnoHardforkDate":"2025-01-21","trinityHardforkDate":"2025-01-21"}
{"level":"info","ts":"2025-03-04T13:11:09.044Z","caller":"sidecar/blockIndexer.go:265","msg":"New tip found, updating","latestTip":21973705,"currentTip":21973674}
{"level":"info","ts":"2025-03-04T13:17:42.600Z","caller":"sidecar/blockIndexer.go:265","msg":"New tip found, updating","latestTip":21973736,"currentTip":21973705}
{"level":"info","ts":"2025-03-04T13:18:00.542Z","caller":"rewards/6_goldRfaeOperators.go:43","msg":"Generating rfae operators table","cutoffDate":"2024-10-14","destTableName":"gold_6_rfae_operators_2024_10_14"}
{"level":"info","ts":"2025-03-04T13:18:07.008Z","caller":"rewards/7_goldActiveODRewards.go:181","msg":"Rewards v2 is not enabled for this cutoff date, skipping Generate7ActiveODRewards"}
{"level":"info","ts":"2025-03-04T13:18:07.008Z","caller":"rewards/8_goldOperatorODRewardAmounts.go:88","msg":"Rewards v2 is not enabled for this cutoff date, skipping GenerateGold8OperatorODRewardAmountsTable"}
{"level":"info","ts":"2025-03-04T13:18:07.008Z","caller":"rewards/9_goldStakerODRewardAmounts.go:132","msg":"Rewards v2 is not enabled for this cutoff date, skipping GenerateGold9StakerODRewardAmountsTable"}
{"level":"info","ts":"2025-03-04T13:18:07.008Z","caller":"rewards/10_goldAvsODRewardAmounts.go:70","msg":"Rewards v2 is not enabled for this cutoff date, skipping GenerateGold10AvsODRewardAmountsTable"}
{"level":"info","ts":"2025-03-04T13:18:07.008Z","caller":"rewards/11_goldStaging.go:133","msg":"Generating gold staging","cutoffDate":"2024-10-14","destTableName":"gold_11_staging_2024_10_14"}
{"level":"info","ts":"2025-03-04T13:18:07.008Z","caller":"rewards/11_goldStaging.go:143","msg":"Is RewardsV2 enabled?","enabled":false}
{"level":"info","ts":"2025-03-04T13:18:31.527Z","caller":"rewards/12_goldFinal.go:32","msg":"Generating gold final table","cutoffDate":"2024-10-14"}
{"level":"info","ts":"2025-03-04T13:19:25.455Z","caller":"stakerOperators/stakerOperator.go:35","msg":"Skipping generation of staker operators table, disabled via config","cutoffDate":"2024-10-14"}
panic: inconsistent label cardinality: expected 0 label values but got 1 in prometheus.Labels{"snapshotDate":"2024-10-14"}

goroutine 129 [running]:
github.com/prometheus/client_golang/prometheus.(*HistogramVec).With(...)
	/go/pkg/mod/github.com/prometheus/client_golang@v1.21.0/prometheus/histogram.go:1273
github.com/Layr-Labs/sidecar/internal/metrics/prometheus.(*PrometheusMetricsClient).Histogram(0xc0006591a0, {0x12dd611, 0x10}, 0x2d59523a740, {0xc004124a80, 0x1, 0x1?})
	/build/internal/metrics/prometheus/prometheus.go:128 +0x232
github.com/Layr-Labs/sidecar/internal/metrics/prometheus.(*PrometheusMetricsClient).Timing(0x20d8fc0?, {0x12dd611?, 0x0?}, 0xc000bc7298?, {0xc004124a80?, 0x1d9422c11f1a8d?, 0xc000bc72f0?})
	/build/internal/metrics/prometheus/prometheus.go:117 +0x1d
github.com/Layr-Labs/sidecar/internal/metrics.(*MetricsSink).Timing(0xc00032a140?, {0x12dd611, 0x10}, 0x2d59523a740, {0xc000bc7350?, 0x1?, 0x12da119?})
	/build/internal/metrics/metrics.go:66 +0x211
github.com/Layr-Labs/sidecar/pkg/rewards.(*RewardsCalculator).calculateRewardsForSnapshotDate.func1()
	/build/pkg/rewards/rewards.go:93 +0xa5
github.com/Layr-Labs/sidecar/pkg/rewards.(*RewardsCalculator).calculateRewardsForSnapshotDate(0xc000331180, {0xc00a923af0, 0xa})
	/build/pkg/rewards/rewards.go:155 +0xc5a
github.com/Layr-Labs/sidecar/pkg/rewards.(*RewardsCalculator).CalculateRewardsForSnapshotDate(...)
	/build/pkg/rewards/rewards.go:159
github.com/Layr-Labs/sidecar/pkg/rewardsCalculatorQueue.(*RewardsCalculatorQueue).processMessage(0xc00068c760, 0xc005cb9350)
	/build/pkg/rewardsCalculatorQueue/handler.go:40 +0x17b
github.com/Layr-Labs/sidecar/pkg/rewardsCalculatorQueue.(*RewardsCalculatorQueue).Process(0xc00068c760)
	/build/pkg/rewardsCalculatorQueue/handler.go:13 +0x197
created by github.com/Layr-Labs/sidecar/cmd.init.func6 in goroutine 1
	/build/cmd/run.go:141 +0x17bb
```

--- a/internal/metrics/metricsTypes/metrics.go
+++ b/internal/metrics/metricsTypes/metrics.go
@@ -51,12 +51,24 @@ var MetricTypes = map[MetricsType][]MetricsTypeConfig{
 			Labels: []string{},
 		},
 		MetricsTypeConfig{
-			Name:   Metric_Incr_GrpcRequest,
-			Labels: []string{},
+			Name: Metric_Incr_GrpcRequest,
+			Labels: []string{
+				"grpc_method",
+				"status",
+				"status_code",
+				"rpc",
+			},
 		},
 		MetricsTypeConfig{
-			Name:   Metric_Incr_HttpRequest,
-			Labels: []string{},
+			Name: Metric_Incr_HttpRequest,
+			Labels: []string{
+				"method",
+				"path",
+				"status_code",
+				"grpc_method",
+				"pattern",
+				"rpc",
+			},
 		},
 	},
 	MetricsType_Gauge: {
@@ -71,24 +83,45 @@ var MetricTypes = map[MetricsType][]MetricsTypeConfig{
 	},
 	MetricsType_Timing: {
 		MetricsTypeConfig{
-			Name:   Metric_Timing_GrpcDuration,
-			Labels: []string{},
+			Name: Metric_Timing_GrpcDuration,
+			Labels: []string{
+				"grpc_method",
+				"status",
+				"status_code",
+				"rpc",
+			},
 		},
 		MetricsTypeConfig{
-			Name:   Metric_Timing_HttpDuration,
-			Labels: []string{},
+			Name: Metric_Timing_HttpDuration,
+			Labels: []string{
+				"method",
+				"path",
+				"status_code",
+				"grpc_method",
+				"pattern",
+				"rpc",
+			},
 		},
 		MetricsTypeConfig{
-			Name:   Metric_Timing_RewardsCalcDuration,
-			Labels: []string{},
+			Name: Metric_Timing_RewardsCalcDuration,
+			Labels: []string{
+				"snapshotDate",
+			},
 		},
 		MetricsTypeConfig{
-			Name:   Metric_Timing_BlockProcessDuration,
-			Labels: []string{},
+			Name: Metric_Timing_BlockProcessDuration,
+			Labels: []string{
+				"rewardsCalculated",
+				"hasError",
+			},
 		},
 		MetricsTypeConfig{
-			Name:   Metric_Timing_CreateSnapshot,
-			Labels: []string{},
+			Name: Metric_Timing_CreateSnapshot,
+			Labels: []string{
+				"chain",
+				"sidecarVersion",
+				"kind",
+			},
 		},
 	},
 }

--- a/internal/metrics/prometheus/prometheus.go
+++ b/internal/metrics/prometheus/prometheus.go
@@ -1,9 +1,12 @@
 package prometheus
 
 import (
+	"fmt"
 	"github.com/Layr-Labs/sidecar/internal/metrics/metricsTypes"
 	"github.com/prometheus/client_golang/prometheus"
 	"go.uber.org/zap"
+	"slices"
+	"strings"
 	"time"
 )
 
@@ -89,6 +92,40 @@ func (pmc *PrometheusMetricsClient) formatLabels(labels []metricsTypes.MetricsLa
 	return l
 }
 
+func (pmc *PrometheusMetricsClient) findExpectedLabels(t metricsTypes.MetricsType, name string) []string {
+	for _, types := range pmc.config.Metrics[t] {
+		if types.Name == name {
+			return types.Labels
+		}
+	}
+	return nil
+}
+
+// hasExpectedLabels checks if any unexpected labels are present in the given labels.
+func (pmc *PrometheusMetricsClient) hasUnexpectedLabels(t metricsTypes.MetricsType, name string, providedLabels []metricsTypes.MetricsLabel) error {
+	if len(providedLabels) == 0 {
+		return fmt.Errorf("no labels provided for metric '%s'", name)
+	}
+	expectedLabels := pmc.findExpectedLabels(t, name)
+	unexpectedLabels := make([]string, 0)
+
+	for _, label := range providedLabels {
+		if !slices.Contains(expectedLabels, label.Name) {
+			unexpectedLabels = append(unexpectedLabels, label.Name)
+		}
+	}
+
+	if len(unexpectedLabels) > 0 {
+		pmc.logger.Sugar().Warnw("Prometheus metric has unexpected labels",
+			zap.String("type", string(t)),
+			zap.String("name", name),
+			zap.Strings("unexpectedLabels", unexpectedLabels),
+		)
+		return fmt.Errorf("unexpected labels: '%s'", strings.Join(unexpectedLabels, ", "))
+	}
+	return nil
+}
+
 func (pmc *PrometheusMetricsClient) Incr(name string, labels []metricsTypes.MetricsLabel, value float64) error {
 	m, ok := pmc.counters[name]
 	if !ok {
@@ -96,6 +133,9 @@ func (pmc *PrometheusMetricsClient) Incr(name string, labels []metricsTypes.Metr
 			zap.String("name", name),
 		)
 		return nil
+	}
+	if err := pmc.hasUnexpectedLabels(metricsTypes.MetricsType_Incr, name, labels); err != nil {
+		return err
 	}
 	m.With(pmc.formatLabels(labels)).Add(value)
 	return nil
@@ -108,6 +148,9 @@ func (pmc *PrometheusMetricsClient) Gauge(name string, value float64, labels []m
 			zap.String("name", name),
 		)
 		return nil
+	}
+	if err := pmc.hasUnexpectedLabels(metricsTypes.MetricsType_Gauge, name, labels); err != nil {
+		return err
 	}
 	m.With(pmc.formatLabels(labels)).Set(value)
 	return nil
@@ -124,6 +167,9 @@ func (pmc *PrometheusMetricsClient) Histogram(name string, value time.Duration, 
 			zap.String("name", name),
 		)
 		return nil
+	}
+	if err := pmc.hasUnexpectedLabels(metricsTypes.MetricsType_Timing, name, labels); err != nil {
+		return err
 	}
 	m.With(pmc.formatLabels(labels)).Observe(float64(value.Milliseconds()))
 	return nil

--- a/internal/metrics/prometheus/prometheus.go
+++ b/internal/metrics/prometheus/prometheus.go
@@ -3,6 +3,7 @@ package prometheus
 import (
 	"fmt"
 	"github.com/Layr-Labs/sidecar/internal/metrics/metricsTypes"
+	"github.com/Layr-Labs/sidecar/pkg/utils"
 	"github.com/prometheus/client_golang/prometheus"
 	"go.uber.org/zap"
 	"slices"
@@ -103,11 +104,19 @@ func (pmc *PrometheusMetricsClient) findExpectedLabels(t metricsTypes.MetricsTyp
 
 // hasExpectedLabels checks if any unexpected labels are present in the given labels.
 func (pmc *PrometheusMetricsClient) hasUnexpectedLabels(t metricsTypes.MetricsType, name string, providedLabels []metricsTypes.MetricsLabel) error {
-	if len(providedLabels) == 0 {
-		return fmt.Errorf("no labels provided for metric '%s'", name)
-	}
 	expectedLabels := pmc.findExpectedLabels(t, name)
 	unexpectedLabels := make([]string, 0)
+
+	if len(expectedLabels) == 0 && len(providedLabels) > 0 {
+		pmc.logger.Sugar().Warnw("Prometheus metric has no expected labels but received labels",
+			zap.String("type", string(t)),
+			zap.String("name", name),
+			zap.Strings("providedLabels", utils.Map(providedLabels, func(label metricsTypes.MetricsLabel, i uint64) string {
+				return label.Name
+			})),
+		)
+		return fmt.Errorf("no expected labels, received '%s'", strings.Join(expectedLabels, ", "))
+	}
 
 	for _, label := range providedLabels {
 		if !slices.Contains(expectedLabels, label.Name) {

--- a/internal/metrics/prometheus/prometheus_test.go
+++ b/internal/metrics/prometheus/prometheus_test.go
@@ -29,12 +29,16 @@ func Test_UnexpectedLabelsParsing(t *testing.T) {
 		})
 		assert.Nil(t, err)
 	})
-	t.Run("Should return an error for no labels", func(t *testing.T) {
-		err := pmc.hasUnexpectedLabels(metricsTypes.MetricsType_Timing, metricsTypes.Metric_Timing_BlockProcessDuration, []metricsTypes.MetricsLabel{})
-		assert.NotNil(t, err)
-	})
 	t.Run("Should return an error for unexpected labels", func(t *testing.T) {
 		err := pmc.hasUnexpectedLabels(metricsTypes.MetricsType_Timing, metricsTypes.Metric_Timing_BlockProcessDuration, []metricsTypes.MetricsLabel{
+			{Name: "rewardsCalculated", Value: "10"},
+			{Name: "hasError", Value: "false"},
+			{Name: "unexpectedLabel", Value: "unexpectedValue"},
+		})
+		assert.NotNil(t, err)
+	})
+	t.Run("Should return an error for unexpected labels when expecting 0 labels", func(t *testing.T) {
+		err := pmc.hasUnexpectedLabels(metricsTypes.MetricsType_Gauge, metricsTypes.Metric_Gauge_CurrentBlockHeight, []metricsTypes.MetricsLabel{
 			{Name: "rewardsCalculated", Value: "10"},
 			{Name: "hasError", Value: "false"},
 			{Name: "unexpectedLabel", Value: "unexpectedValue"},

--- a/internal/metrics/prometheus/prometheus_test.go
+++ b/internal/metrics/prometheus/prometheus_test.go
@@ -1,0 +1,45 @@
+package prometheus
+
+import (
+	"github.com/Layr-Labs/sidecar/internal/logger"
+	"github.com/Layr-Labs/sidecar/internal/metrics/metricsTypes"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func Test_UnexpectedLabelsParsing(t *testing.T) {
+	l, err := logger.NewLogger(&logger.LoggerConfig{Debug: false})
+	assert.Nil(t, err)
+
+	pmc, err := NewPrometheusMetricsClient(&PrometheusMetricsConfig{
+		Metrics: metricsTypes.MetricTypes,
+	}, l)
+	assert.Nil(t, err)
+
+	t.Run("Should return no error for all labels", func(t *testing.T) {
+		err := pmc.hasUnexpectedLabels(metricsTypes.MetricsType_Timing, metricsTypes.Metric_Timing_BlockProcessDuration, []metricsTypes.MetricsLabel{
+			{Name: "rewardsCalculated", Value: "10"},
+			{Name: "hasError", Value: "false"},
+		})
+		assert.Nil(t, err)
+	})
+	t.Run("Should return no error for a subset labels", func(t *testing.T) {
+		err := pmc.hasUnexpectedLabels(metricsTypes.MetricsType_Timing, metricsTypes.Metric_Timing_BlockProcessDuration, []metricsTypes.MetricsLabel{
+			{Name: "rewardsCalculated", Value: "10"},
+		})
+		assert.Nil(t, err)
+	})
+	t.Run("Should return an error for no labels", func(t *testing.T) {
+		err := pmc.hasUnexpectedLabels(metricsTypes.MetricsType_Timing, metricsTypes.Metric_Timing_BlockProcessDuration, []metricsTypes.MetricsLabel{})
+		assert.NotNil(t, err)
+	})
+	t.Run("Should return an error for unexpected labels", func(t *testing.T) {
+		err := pmc.hasUnexpectedLabels(metricsTypes.MetricsType_Timing, metricsTypes.Metric_Timing_BlockProcessDuration, []metricsTypes.MetricsLabel{
+			{Name: "rewardsCalculated", Value: "10"},
+			{Name: "hasError", Value: "false"},
+			{Name: "unexpectedLabel", Value: "unexpectedValue"},
+		})
+		assert.NotNil(t, err)
+	})
+
+}

